core/config/config.go

```diff
@@ -144,41 +145,61 @@ func GetEnvironmentPrefix() string {
 // ProviderFunc is used to create config providers on configuration initialization
 type ProviderFunc func() (ConfigurationProvider, error)
 
+// DynamicProviderFunc is used to create config providers on configuration initialization
```

Comment:
We should probably mention why this provider takes a provider as an argument :)

Something along the lines of: "dynamic provider configuration settings (ports, keys, etc) can be supplied via env or yaml"